src/crypto/equihash.cpp

```diff
@@ -0,0 +1,287 @@
+{
+    unsigned char personalization[crypto_generichash_blake2b_PERSONALBYTES] = {};
+    memcpy(personalization, "ZcashPOW", 8);
+    memcpy(personalization+8,  &n, 4);
+    memcpy(personalization+12, &k, 4);
+    return crypto_generichash_blake2b_init_salt_personal(&base_state,
+                                                         NULL, 0, // No key.
+                                                         n/8,
+                                                         NULL,    // No salt.
+                                                         personalization);
+}
+
+StepRow::StepRow(unsigned int n, const eh_HashState& base_state, eh_index i) :
+        hash {new unsigned char[n/8]},
+        len {n/8},
+        indices {i}
+{
+    eh_HashState state;
+    state = base_state;
+    crypto_generichash_blake2b_update(&state, (unsigned char*) &i, sizeof(eh_index));
+    crypto_generichash_blake2b_final(&state, hash, n/8);
+
+    assert(indices.size() == 1);
+}
+
+StepRow::~StepRow()
+{
+    delete[] hash;
+}
+
+StepRow::StepRow(const StepRow& a) :
+        hash {new unsigned char[a.len]},
+        len {a.len},
+        indices(a.indices)
+{
+    for (int i = 0; i < len; i++)
+        hash[i] = a.hash[i];
+}
+
+StepRow& StepRow::operator=(const StepRow& a)
+{
+    unsigned char* p = new unsigned char[a.len];
+    for (int i = 0; i < a.len; i++)
+        p[i] = a.hash[i];
+    delete[] hash;
+    hash = p;
+    len = a.len;
+    indices = a.indices;
+    return *this;
+}
+
+StepRow& StepRow::operator^=(const StepRow& a)
+{
+    if (a.len != len) {
+        throw std::invalid_argument("Hash length differs");
+    }
+    if (a.indices.size() != indices.size()) {
+        throw std::invalid_argument("Number of indices differs");
+    }
+    unsigned char* p = new unsigned char[len];
+    for (int i = 0; i < len; i++)
+        p[i] = hash[i] ^ a.hash[i];
+    delete[] hash;
+    hash = p;
+    indices.reserve(indices.size() + a.indices.size());
+    indices.insert(indices.end(), a.indices.begin(), a.indices.end());
+    return *this;
+}
+
+void StepRow::TrimHash(int l)
+{
+    unsigned char* p = new unsigned char[len-l];
+    for (int i = 0; i < len-l; i++)
+        p[i] = hash[i+l];
+    delete[] hash;
+    hash = p;
+    len -= l;
+}
+
+bool StepRow::IsZero()
+{
+    char res = 0;
+    for (int i = 0; i < len; i++)
+        res |= hash[i];
+    return res == 0;
+}
+
+bool HasCollision(StepRow& a, StepRow& b, int l)
+{
+    bool res = true;
+    for (int j = 0; j < l; j++)
+        res &= a.hash[j] == b.hash[j];
+    return res;
+}
+
+// Checks if the intersection of a.indices and b.indices is empty
+bool DistinctIndices(const StepRow& a, const StepRow& b)
+{
+    std::vector<eh_index> aSrt(a.indices);
+    std::vector<eh_index> bSrt(b.indices);
+
+    std::sort(aSrt.begin(), aSrt.end());
+    std::sort(bSrt.begin(), bSrt.end());
+
+    unsigned int i = 0;
+    for (unsigned int j = 0; j < bSrt.size(); j++) {
+        while (aSrt[i] < bSrt[j]) {
```

Comment:
This is a bounds error if ``aSrt`` has a shorter length than ``bSrt``. Let's either check for proper input or document a precondition.